plutus-playground-client/src/Editor/State.purs

```diff
@@ -2,43 +2,62 @@ module Editor.State where
 
 import Editor.Types
 import Control.Alternative ((<|>))
-import Data.Maybe (Maybe, fromMaybe, maybe)
+import Data.Lens (assign, modifying, use)
+import Data.Maybe (Maybe(..), fromMaybe, maybe)
 import Effect.Aff.Class (class MonadAff, liftAff)
 import Effect.Class (class MonadEffect)
 import FileEvents (preventDefault, readFileFromDragEvent)
 import Halogen (HalogenM, liftEffect, query, tell)
 import Halogen.Monaco (KeyBindings(..))
 import Halogen.Monaco (Message(..), Query(..)) as Monaco
+import Language.Haskell.Interpreter (SourceCode(SourceCode))
 import LocalStorage (Key)
 import LocalStorage as LocalStorage
 import Monaco (Editor, getModel, layout, focus, setPosition, setValue) as Monaco
-import Prelude (Unit, bind, discard, pure, show, unit, void, ($), (<$>))
+import Prelude (Unit, bind, discard, not, pure, show, unit, void, ($), (<$>), (==))
 import Types (ChildSlots, _editorSlot)
 
 initialState :: forall m. MonadEffect m => m State
 initialState =
   liftEffect do
     keyBindings <- loadKeyBindings
-    pure $ State { keyBindings }
+    pure
+      $ State
+          { keyBindings
+          , feedbackPaneMinimised: false
+          , lastCompiledCode: Nothing
+          , currentCodeIsCompiled: false
+          }
 
 handleAction ::
-  forall state action output m.
+  forall action output m.
   MonadEffect m =>
   MonadAff m =>
   Key ->
   Action ->
-  HalogenM state action ChildSlots output m Unit
+  HalogenM State action ChildSlots output m Unit
 handleAction bufferLocalStorageKey Init = do
   binding <- loadKeyBindings
+  assign _keyBindings binding
   handleAction bufferLocalStorageKey (SetKeyBindings binding)
 
-handleAction bufferLocalStorageKey (HandleEditorMessage (Monaco.TextChanged text)) = liftEffect $ saveBuffer bufferLocalStorageKey text
+handleAction bufferLocalStorageKey (HandleEditorMessage (Monaco.TextChanged text)) = do
+  lastCompiledCode <- use _lastCompiledCode
+  case lastCompiledCode of
+    Just (SourceCode code) -> assign _currentCodeIsCompiled (code == text)
+    Nothing -> assign _currentCodeIsCompiled false
+  liftEffect $ saveBuffer bufferLocalStorageKey text
 
 handleAction _ (SetKeyBindings binding) = do
   void $ query _editorSlot unit $ tell $ Monaco.SetKeyBindings binding
   void $ query _editorSlot unit $ tell $ Monaco.Focus
+  void $ query _editorSlot unit $ tell $ Monaco.Resize
```

Comment:
I feel like we should have our own `tell slot action = void $ query slot unit $ tell action` but I guess the naming might be confusing. It's just that this is so common and it looks messy.

Reply:
Hmm...looking at it I wonder if we can do something like:

``` purescript
handleAction _ (SetKeyBindings binding) = void $ query _editorSlot unit $ do
  tell $ Monaco.SetKeyBindings binding
  tell Monaco.Focus
  tell Monaco.Resize
```

Reply:
I agree this repetition should be avoided. Not immediately sure how best to do it, though, and it doesn't matter for the refresh. So unless inspiration strikes, I'm inclined to leave it as it is for now.

Reply:
I agree.